lib/min_heap.rb

```diff
@@ -44,28 +54,53 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(1)
+  # Space complexity: O(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    return true if @store.length == 0
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(log n) only comparing with its parent on one side of the tree
+  # Space complexity: O(log n) stack trace of heap_up and swap
   def heap_up(index)
-    
+    return if index == 0
+
+    # compares with its parent element - if smaller, swap
+    parent_index = (index - 1) / 2
+
+    if @store[index].key < @store[parent_index].key
+      swap(index, parent_index)
+      heap_up(parent_index)
+    end
   end
 
-  # This helper method takes an index and 
-  #  moves it up the heap if it's smaller
-  #  than it's parent node.
   def heap_down(index)
```

Comment:
This is passing the tests, but I think there may be a subtle error here that can come in a few circumstances.